ext/hal/st/stm32cube/stm32f1xx/drivers/src/stm32f1xx_ll_utils.c

```diff
@@ -559,15 +559,15 @@ static ErrorStatus UTILS_EnablePLLAndSwitchSystem(uint32_t SYSCLK_Frequency, LL_
   /* Update system clock configuration */
   if (status == SUCCESS)
   {
-#if defined(RCC_PLL2_SUPPORT)
+#if defined(RCC_PLL2_SUPPORT) && !defined(CONFIG_CLOCK_STM32_PLL_SRC_HSI)
```

Comment:
Ideally we shouldn't modify hal with Zephyr code.
Isn't there a non zephyr specific way to fix this?

Reply:
I think in `ext/hal/st/stm32cube/stm32f1xx/CMakeLists.txt` you should be able to unset it.
Please document the README, I'll raise an issue in HAL, I'll give you the tracking number

Reply:
tracking number: 51414

Reply:
That doesn't seem to work. It doesn't surprise me since `RCC_PLL2_SUPPORT` is not a config option but a `#define`.

Reply:
Tried `zephyr_cc_option_ifdef(CONFIG_CLOCK_STM32_PLL_SRC_HSI -URCC_PLL2_SUPPORT)` but that also doesn't seem to stick...

Reply:
Sure. Anyway, activating HSI as PLL source does not mean PLL2 is not supported, so this was not ideal, even as temp fix.
Then, can you have a try a cube compatible way to solve this with for instance a parameter to the function ?

Reply:
> Sure. Anyway, activating HSI as PLL source does not mean PLL2 is not supported, so this was not ideal, even as temp fix.

Fair enough. However this function deals with setting up the SysClock and you can only use PLL2 for the SysClock if you use HSE. I don't think the HAL should bother with a potential parallel use of HSE and PLL2 for MCO in this function.

> Then, can you have a try a cube compatible way to solve this with for instance a parameter to the function ?

How would that work? The block I'm trying to gate is trying to turn on PLL2 unconditionally and then waits for it to come up (which will never happen so that's a deadlock). The only "compatible" fix would be not to use this function at all and instead initialise it "manually" in the clock driver, like CubeMX does.

Reply:
@therealprof, we should find a way to fix this as it would be fixed inside a standalone Cube using LL api.
In the end if this is an actual problem, there should be a way to solve it outside of Zephyr and w/o use of zephyr CONFIG_.
Can't we check for clock source set in registers and act upon this information?

Reply:
> In the end if this is an actual problem

Well, if you consider the device just hanging in an endless loop during init, then this would be an actual problem.

The only alternative fixes outside of the LL API I see are:
* Not using this magic do-it-all utility function to do the clock initialisation (like the CubeMX version)
* Somehow undefining RCC_PLL2_SUPPORT

> Can't we check for clock source set in registers and act upon this information?

How would that work? If we use `UTILS_EnablePLLAndSwitchSystem()` we will run in this problem (unless we can get rid of the define) -- always. Even if there was a way around it by checking registers (no idea at the moment whether that's a real possibility), if this function is used it's game over.

Luckily there's another severe problem with the HAL for the STM32F107 somewhere which I couldn't localise yet so for me this has no priority at the moment... Maybe it'll get fixed upstream some time.

Reply:
My point was to use `LL_RCC_PLL_GetMainSource` in  `UTILS_EnablePLLAndSwitchSystem()` to do this check before enabling PLL2.
Can't  it work?

Reply:
@therealprof ,

Coming back on this. Here is my proposal, which I'm unfortunately unable to test.
To be clear my proposal is to modify `UTILS_EnablePLLAndSwitchSystem()` using cube function, so in a way it could be maintained outside of zephyr.

```suggestion
#if defined(RCC_PLL2_SUPPORT)
-    /* Enable PLL2 */
-    LL_RCC_PLL2_Enable();
-    while (LL_RCC_PLL2_IsReady() != 1U)
-    {
-     /* Wait for PLL2 ready */
+    /* Wait PLL2 only if set as PLL source */
+    if (LL_RCC_PLL_GetMainSource() == LL_RCC_PLLSOURCE_PLL2) {
+	    /* Enable PLL2 */
+	    LL_RCC_PLL2_Enable();
+	    while (LL_RCC_PLL2_IsReady() != 1U)
+	    {
+	      /* Wait for PLL2 ready */
+	    }
```

Reply:
Hm, I'd have to check that possibility but I don't know where my (orphaned) prototype went; might have to build another one but I'm lacking the time for that at the moment, I'm afraid.

Reply:
Ok. Thanks for the answer anyway. I'll raise an internal issue